config_utilities/src/external_registry.cpp

```diff
@@ -0,0 +1,217 @@
+    // These modes are chosen to somewhat closely mimic dlopen. It will both append appropriate prefixes and suffixes
+    // (lib and .so for linux) and allows loading from LD_LIBRARY_PATH by just specfiying the library name.
+    // Your mileage may vary on windows, though it looks like boost::dll does a good job picking defaults that
+    // are equivalent between the platforms.
+    const auto mode = boost::dll::load_mode::append_decorations | boost::dll::load_mode::search_system_folders;
+    library = boost::dll::shared_library(library_path.string(), mode);
+  }
+
+  boost::dll::shared_library library;
+};
+
+LibraryGuard::LibraryGuard() {}
+
+LibraryGuard::LibraryGuard(const std::filesystem::path& library) { libraries_.push_back(library); }
+
+LibraryGuard::~LibraryGuard() { unload(); }
+
+LibraryGuard::LibraryGuard(LibraryGuard&& other) {
+  libraries_.insert(libraries_.end(), other.libraries_.begin(), other.libraries_.end());
+  other.libraries_.clear();
+}
+
+LibraryGuard& LibraryGuard::operator=(LibraryGuard&& other) {
+  libraries_.insert(libraries_.end(), other.libraries_.begin(), other.libraries_.end());
+  other.libraries_.clear();
+  return *this;
+}
+
+void LibraryGuard::unload() {
+  for (const auto& library : libraries_) {
+    ExternalRegistry::instance().unload(library);
+  }
+
+  libraries_.clear();
+}
+
+LibraryGuard::operator bool() const { return !libraries_.empty(); }
+
+ExternalRegistry::~ExternalRegistry() {
+  std::vector<std::string> libraries;
+  for (const auto& [path, lib] : libraries_) {
+    libraries.push_back(path);
+  }
+
+  for (const auto& path : libraries) {
+    unload(path);
+  }
+}
+
+void ExternalRegistry::unload(const std::filesystem::path& library_path) {
+  if (Settings::instance().verbose_external_load) {
+    // NOTE(nathan) this is separate from the logger becuase there is no guarantee that it will be visible to the user
+    // if it is through the logger
+    std::cerr << "[WARNING] Unloading external library: " << library_path << std::endl;
+  }
+
+  auto iter = entries_.find(library_path);
+  if (iter != entries_.end()) {
+    // Remove any factories that use code from the external library in question. This will also delete any factory
+    // module maps that become empty (as they rely on types only defined in the plugin). While plugin libraries
+    // shouldn't do this, it's easy to happen in practice and will cause segfaults
+    for (const auto& entry : iter->second) {
+      internal::ModuleRegistry::removeModule(entry.key, entry.type);
+      all_entries_.erase(entry);
+    }
+  }
+
+  for (const auto& internal_ref : instances_) {
+    const auto actual_ref = internal_ref.lock();
+    if (actual_ref) {
+      actual_ref->cleanup();
+    }
+  }
+
+  entries_.erase(library_path);
+  libraries_.erase(library_path);
+  instances_.clear();
+}
+
+LibraryGuard ExternalRegistry::load(const std::filesystem::path& library_path) {
```

Comment:
Would it make sense / is it feasible to check whether the library to be loaded exists and is found by LD_PATH? That would allow importing libraries on a 'as available' basis, or warn the user why the library could not be found?

Reply:
It might have to be done by a `try / catch` instead of preemptively checking, but yeah that's a good point (and I should actually be handling `boost` failing here)